app/code/Magento/Checkout/Model/ShippingInformationManagement.php

```diff
@@ -199,26 +214,39 @@ public function saveAddressInformation(
             );
         }
 
-        /** @var \Magento\Checkout\Api\Data\PaymentDetailsInterface $paymentDetails */
+        /** @var PaymentDetailsInterface $paymentDetails */
         $paymentDetails = $this->paymentDetailsFactory->create();
         $paymentDetails->setPaymentMethods($this->paymentMethodManagement->getList($cartId));
         $paymentDetails->setTotals($this->cartTotalsRepository->get($cartId));
         return $paymentDetails;
     }
 
+    /**
+     * Validate shipping address
+     *
+     * @param AddressInterface $address
+     * @return void
+     * @throws StateException
+     */
+    private function validateAddress(AddressInterface $address): void
+    {
+        if (!$address || !$address->getCountryId()) {
+            throw new StateException(__('The shipping address is missing. Set the address and try again.'));
+        }
+    }
+
     /**
      * Validate quote
      *
-     * @param \Magento\Quote\Model\Quote $quote
+     * @param Quote $quote
      * @throws InputException
-     * @throws NoSuchEntityException
      * @return void
      */
-    protected function validateQuote(\Magento\Quote\Model\Quote $quote)
+    protected function validateQuote(Quote $quote): void
     {
-        if (0 == $quote->getItemsCount()) {
+        if (!$quote->getItemsCount()) {
```

Comment:
Actually `0 === $quote->getItemsCount()` might be more efficient, as it does not juggle the type.